### PR TITLE
Center videos correctly & fix padding for controls

### LINKED ISSUE #4040: Content not centered with 18:9 screen -> weird behaviour with navigation bar
### Version

- d8b5549fd98e43f18b993c5f36f2385cf9a80620 

### Steps to reproduce the bug
1. Look for a vertical video, eg [this one](https://www.youtube.com/watch?v=KRntP-q_R9s)
2. Launch fullscreen

### Expected behavior

It is centered, and there the video doesn't "jump" to give space for navigation bar
### Actual behaviour

It's not centered, then when you click, the video goes up so that there is space for navigation bar
### Screenshots/Screen recordings

0.19.8 
0 19 8_nav_bar_1 0 19 8_nav_bar_2

unified
unified_nav_bar_1 unified_nav_bar_2

gif for this "jump" with unified

### PR DESCRIPTION
**Pull request by blackbox87**
## What is it?
- [x] Bug fix (user facing)
- [ ] Feature (user facing)
- [ ] Code base improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- Centers vertical and landscape videos correctly on devices with cutouts
- Prevents the controls from being placed under corner cutouts by adding a little extra padding while in landscape

#### Fixes the following issue(s)
It improves on #4154 and fully fixes #4040

#### Testing apk
[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5196395/app-debug.zip)

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

### CODE REVIEW

**Comment by Domiuns** [comment]
@blackbox87 this fix also this https://github.com/TeamNewPipe/NewPipe/issues/4179 ?

**Comment by blackbox87** [comment]
@Domiuns On my device it only takes me 1 tap to see both the status bar and the video controls and this commit does nothing to change that.

**Comment by blackbox87** [comment]
@Domiuns On my device it only takes me 1 tap to see both the status bar and the video controls and this commit does nothing to change that.

**Comment by blackbox87** [comment]
Sorry about the forced pushes. I messed up and needed to correct things.

> * Prevents the controls from being placed under corner cutout by adding a little extra padding while in landscape

Basically it's difficult to accurately detect if you've got a cutout, it's position and it's size. So without the padding you might end up having the controls or the videos title placed under a cutout. That's why padding is now being applied to all devices as a precaution.

I've tested the changes on my old phone without a cutout and my new phone with a cutout. It looks good to me.

**Comment by blackbox87** [comment]
Sorry about the forced pushes. I messed up and needed to correct things.

> * Prevents the controls from being placed under corner cutouts by adding a little extra padding while in landscape

Basically it's difficult to accurately detect if you've got a cutout, it's position and it's size. So without the padding you might end up having the controls or the videos title placed under a cutout. That's why padding is now being applied to all devices as a precaution.

I've tested the changes on my old phone without a cutout and my new phone with a cutout. It looks good to me.

**Comment by opusforlife2** [comment]
Could upload a test apk so other people can test it?

**Comment by wb9688** [comment]
@blackbox87: Can't you use https://developer.android.com/reference/android/view/WindowInsets#getDisplayCutout()?

Btw force pushing isn't a problem for PRs, we like having a clean (but not "too clean") commit history.

Also, do you post, then delete and repost a comment? I get multiple notifications from your comments, but if that's the case, could you please just edit your comment?

**Comment by blackbox87** [comment]
@opusforlife2 I'll do that shortly. I'm currently away from my PC.

@wb9688 Not easily, no. Mainly because you want the controls to always be centered on the screen, but when using that API you'll need to calculate things to center the controls and it needs to change as you rotate your device.

I opted to do it the easier way, which is to add enough padding to the left and right sides of the screen so that the controls are centered. And although it's possible not to apply that extra padding on devices without a cutout I haven't done that for the sake of consistency.

> Also, do you post, then delete and repost a comment? I get multiple notifications from your comments, but if that's the case, could you please just edit your comment?

I don't believe I have, at least not deliberately.

**Comment by avently** [comment]
@blackbox87 I thought extra padding will look bad but actually it is quite nice with your code. I didn't tested everything, just launched the app to see the changes of look.

The only thing I want to ask you is to remove extra padding for devices without cutout and here is why. I don't want to pay for someone's ugly displays. Yes, I hate these big terrible cutouts and never liked it. You or someone else may disagree but if you like it, not a problem, just don't add extra padding for those who has displays without cutouts. 

Big thank you that you found a way to work around this big Android API problem.

**Comment by avently** [comment]
Found an issue while testing. 
To reproduce:
Use vertical screen orientation
Open vertical video (billie eilish - bad guy, for example, our classic :))
Add any non-vertical video to queue (if you have a checkbox Autoplay checked, it will be added automatically). To make this just long press on a video from related videos and choose Enqueue in background
Press on thumbnail to start playing in main player
Click on fullscreen button to enlarge the video
Click on next video arrow button

You'll see that padding on top and bottom is removed, so controls located under navigation bar, and status bar, and cutout.

Let me know if I need to make a screenrecord.

**Comment by blackbox87** [comment]
@avently 
> Let me know if I need to make a screenrecord.

Please do as I'm probably doing something wrong, but I can't replicate that issue. Seeing it would help.

I'm probably probably going to adjust the code slightly anyway since I just tested it on Android 4.4 emulator and I don't like how the black navbar makes the controls look off center, even though they are centered.

**Comment by opusforlife2** [comment]
> Open vertical video (billie eilish - bad guy, for example, our classic :))

You mean "you should see me in a crown" :P

**Comment by avently** [comment]
[ScreenRecord-2020-09-09-20-57-21.zip](https://github.com/TeamNewPipe/NewPipe/files/5197120/ScreenRecord-2020-09-09-20-57-21.zip)

**Comment by blackbox87** [comment]
@avently 
So it looks like setControlsSize() should be called when switching between videos. Most of this code is yours, so any suggestions where the best place to put that is?

**Comment by avently** [comment]
> So it looks like setControlsSize() should be called when switching between videos.

Better to avoid it because you can call this in the wrong place (while video is rotating, for example). And the resulting values for screen elements will be wrong. 
In the video you may see that padding was changed when video was changed, it means that you need to find a place in code where such change was happen and to adjust the code based on new information. 

setControlsSize() is kind of magic. I do here what Android should do for us, so it's hard to say to you the correct place when the code should be changed and how. But the main rule is to avoid calling this method many times because it can be called in many device's UI states

**Comment by avently** [comment]
Maybe this is a problem of the issue: https://github.com/TeamNewPipe/NewPipe/blob/6b27bce7692fb8b380d8d6c48b5979900070e425/app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java#L1542

Here you don't think about situation when a video can be in fullscreen but not in landscape. But it's looking at the code on github, didn't tested.

**Comment by avently** [comment]
Looks like some day we will end up by removing all this custom code in favor of UI without status bar and navigation bar that's simple but not so useful:) But for now let's do cool custom things:)

**Comment by opusforlife2** [comment]
I have no cutouts. In a vertical video (we all know which one :P ) the bottom stuff (seekbar etc.) is padded, but the top stuff isn't.

But I agree with avently that devices without cutouts shouldn't be affected by this PR at all.

**Comment by MD77MD** [comment]
noooo @avently, anything but the status bar. like you said, let's stay cool 😎 and useful. 😊

**Comment by Domiuns** [comment]
Nice, this apk crash and block my device

**Comment by blackbox87** [comment]
@Domiuns None of my changes should cause that, so maybe the issue is with the current dev branch.

Instead of removing your battery you can probably hold your power button down for 10-20 seconds.

**Comment by Domiuns** [comment]
@blackbox87 demostration https://streamable.com/e/ygkbwt

**Comment by blackbox87** [comment]
@Domiuns Try the 2020-09-08 version posted here > https://github.com/TeamNewPipe/NewPipe/projects/17#card-42918515

Does it have the same issue?

@avently I fixed the bug that you reported, but now I'm stuck trying to detect cutouts correctly. Your hasCutout() method doesn't work because it assumes that the status bar height will be adjusted for a cutout, which isn't the case on my device. So I've tried using getDisplayCutout() instead, but if I start NewPipe in landscape then it'll often return null.

**Comment by avently** [comment]
@blackbox87 can't help you in this case because never did something with cutouts in my apps. If you will not find a solution, not a problem, will leave as the current state of dev branch. Not a big deal I think to have a non-centered vertical video and masked cutout space in landscape.

**Comment by blackbox87** [comment]
> Not a big deal I think to have a non-centered vertical video and masked cutout space in landscape.

I think it is a problem since it worked perfectly before the unified player was added. That's the only reason I'm still trying to fix this.

At the moment I can't say that I'm a fan of the unified player as it's introduced more issues than it solves.

**Comment by avently** [comment]
@blackbox87 
> At the moment I can't say that I'm a fan of the unified player as it's introduced more issues than it solves.

Maybe you right, maybe not. The only I can say that I never used NewPipe without unified player because the app was unusable. It doesn't mean it was bad, no, it was awesome but uncomfortable. So you may be a fan or not but now you have more comfort than before. 

>That's the only reason I'm still trying to fix this.

The only reason you still trying is that you, like me, want to use the app in it's ideal state of look and feel. And that's great! You are making it better with every commit

**Comment by opusforlife2** [comment]
> The only reason you still trying is that you, like me, want to use the app in it's ideal state of look and feel. And that's great! You are making it better with every commit

Love this line. <3

**Comment by Domiuns** [comment]
@blackbox87 you're right the problem is not caused by your PR but by what was merged on apk version 530f745 2020-09-08

**Comment by avently** [comment]
@Domiuns 
Maybe you'll have more luck with this one? https://github.com/TeamNewPipe/NewPipe/issues/4230#issuecomment-690686697

I don't think it's related to commit you linked because nothing serious was made in it. Maybe in previous PRs...

**Comment by blackbox87** [comment]
Here's a new test version. It should now detect cutouts correctly and won't apply extra padding unless it's necessary.

If nobody can find any issues then I'll update this pull request with the changes.

@avently This should also fix the bug you reported.

[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5211554/app-debug.zip)

**Comment by opusforlife2** [comment]
No extra padding showing for me when comparing this and the latest unified apk. 👍 

Does the seekbar jump up a little (just a pixel or two) when switching from another app to a fullscreen video for anyone else? This happens in the latest unified apk as well.

**Comment by blackbox87** [comment]
> Does the seekbar jump up a little (just a pixel or two) when switching from another app to a fullscreen video for anyone else? This happens in the latest unified apk as well.

It looks fine on my devices, but I have seen the seekbar look off-placed once when trying NewPipe on an emulator. It seems to be caused by the changes someone made to the seekbar.

**Comment by avently** [comment]
@opusforlife2 try this PR maybe with a fix for seekbar issue #4272

**Comment by avently** [comment]
@blackbox87 
>This should also fix the bug you reported.

You just dropped fullscreen when moving from vertical video in portrait + fullscreen into non-vertical video. Smart choice!

Overall the app works fantastic. No issues I found expect for double cutout (`more button` located under navigation bar a little bit). If you don't know how to fix it, no problem. I don't think that there are many phones with this double cutout thing.

Great job!

**Comment by blackbox87** [comment]
@avently I could try to fix that if I could emulate it. I guess you need to be using Android 10 or 11 with the gesture navbar?

**Comment by avently** [comment]
No, android 9 with two buttons navbar and double cutout from developer options.
S
To reproduce open any non-vertical video in landscape with double cutout enabled. Try both sides of device because in one side its ok, but not in another

**Comment by ZeroPucks** [comment]
Unsure if this is correct place to report this, but testing latest APKs from #3178  , #4222 , and #4255  on Android 11 with Pixel 3a (2220x1080 resolution, no cutouts) and watching a landscape video fullscreen results in a persistent red status bar and no seekbar, as shown below.

Screenshot_20200913-100018_1
Screenshot_20200913-100035_1

Below is a screenshot from 0.19.8 APK:
Screenshot_20200913-114858_1

Just a bummer while on Android 11 because users may have to choose between having correct video playback but broken background notifications, or correct background notifications (from #3178 ) but broken video playback. Let me know if any logs are needed. Thanks to all contributors!

**Comment by avently** [comment]
@ZeroPucks open new issue, please. This is not correct place for it because it's unrelated to this PR.

**Comment by blackbox87** [comment]
@avently Fixed it.

[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5214717/app-debug.zip)

I've also tested the changes against Android 10's gesture navigation.

**Comment by avently** [comment]
@blackbox87 cool! Thank you. So you're ready to upload the latest changes and to merge the PR? Or something is not ready yet?

**Comment by blackbox87** [comment]
If everything looks good and nobody can find any issues then I'll update this PR and then it should be ready for merging.

I'm currently waiting to see if my latest changes also fix #4277.

**Comment by avently** [comment]
Based on my testing, all works perfectly:
- phone without cutout, single cutout, double cutout
- tablet without cutout
- all orientations
- multiwindow mode
- vertical videos with single cutout and without (didn't tested other variants).

So, yeah, works great. Great job, man!

**Comment by avently** [comment]
But note that I use a phone with 16:9 aspect ratio, on other screens situation maybe different

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java [hunk 1]
```diff
@@ -203,6 +207,11 @@
     private int cachedDuration;
     private String cachedDurationString;
 
+    private boolean gotCutout;
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java:210]:
```suggestion
    private boolean hasCutout;
```

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java [hunk 2]
```diff
@@ -1508,7 +1517,84 @@ public void setControlsSize() {
...
+            } else {
+                if (gotCutout) {
+                    ctrlPadLeft = controlsPadding
+                            + (getNavBarMode() == 2 ? 0 : navBarHeight) + insetLeft;
+                    ctrlPadRight = controlsPadding + insetRight;
+                } else {
+                    ctrlPadLeft = controlsPadding;
+                    ctrlPadRight = controlsPadding;
+                }
+            }
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java:1566]: why do you only check for 90 deg, but not for 180?

**Review by TobiGr — COMMENTED** [review]
thank you for the effort!

**Comment by avently** [comment]
@TobiGr there is more recent method of doing the same thing as this PR but in more correct/universal way: https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-69267491

This PR contains one more thing (useful for @blackbox87) that can be ported over my PR: more padding for devices with hidden navbar and enabled gestures on custom ROMs. We have a talk related to all this in the link I wrote.

**Comment by blackbox87** [comment]
@TobiGr When it's 90° in landscape then the navbar is on the right side of your screen, so the else statement handles the opposite. And this orientation check only applies to phones in landscape, so it always seems to work.

If you look at #4272 then you'll see that avently worked on a cleaner method, but we had a disagreement about handling a navbar feature that's found on a lot of custom Android 10 ROMs (LineageOS, ArrowOS, Resurrection Remix, PixelExperience, AICP etc).

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java (hunk 2 above)
Comment (blackbox87) [reply app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java:1566]: @TobiGr When it's 90° in landscape then the navbar is on the right side of your screen, so the else statement handles the opposite. And this orientation check only applies to phones in landscape, so it always seems to work.

If you look at #4272 then you'll see that avently worked on a cleaner method, but we had a disagreement about handling a navbar feature that's found on a lot of custom Android 10 ROMs (LineageOS, ArrowOS, Resurrection Remix, PixelExperience, AICP etc).

**Comment by blackbox87** [comment]
@TobiGr @Stypox I'm closing this PR because #4272 handles the insets in a cleaner way, although I still think [the overlapped gesture navbar issue](https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-693111775) should be handled by #4272 rather than a separate PR. And I'm sure Avently will disagree, but the only reason I opened this PR was because he didn't want to fix the issue until I showed that it could still be done. So originally I had to base my changes on top of his, then rebase to the dev branch and then Avently looked at my changes and bundled his own version of the cutout fix into a PR that fixes multiple different issues.

I won't submit another PR with a fix for hidden gesture navbars, but if someone else wants to do it then [here's a fix that's been tested against a few Android 10 ROMs](https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-693124017).

**Comment by fareszr** [comment]
i can confirm this works with a galaxy s10+ with modpunk Linage os.

**Comment by fareszr** [comment]
> @TobiGr @Stypox I'm closing this PR because #4272 handles the insets in a cleaner way, although I still think [the overlapped gesture navbar issue](https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-693111775) should be handled by #4272 rather than a separate PR. And I'm sure Avently will disagree, but the only reason I opened this PR was because he didn't want to fix the issue until I showed that it could still be done. So originally I had to base my changes on top of his, then rebase to the dev branch and then Avently looked at my changes and bundled his own version of the cutout fix into a PR that fixes multiple different issues.
> 
> I won't submit another PR with a fix for hidden gesture navbars, but if someone else wants to do it then [here's a fix that's been tested against a few Android 10 ROMs](https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-693124017).

#4272 doesn't fix the black notification bar problem.

**Comment by avently** [comment]
@blackbox87 thank you for your contribution! Without it I wouldn't start thinking about cleaner and better way of positioning the player. Your PR is really important for me because there are a few people who really do PRs here. And I was surprised that you started to to so. Even if we disagree with each other in some situations I think we did a great job and can do it in the future too.

**Comment by blackbox87** [comment]
@fareszr Black notification bar problem? I'm not sure which that is, but I'd guess that once #3178, #4272 and #4288 are merged then it'll be fixed.

You could try using this(https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-695350659) or this(https://github.com/TeamNewPipe/NewPipe/pull/4288#issuecomment-697097581)? But you'll probably want to continue that discussion in #4272 if the issue occurs because of one of avently's commits.

**Comment by fareszr** [comment]
> @fareszr Black notification bar problem? I'm not sure which that is, but I'd guess that once #3178, #4272 and #4288 are merged then it'll be fixed.
> 
> You could try using this([#4272 (comment)](https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-695350659)) or this([#4288 (comment)](https://github.com/TeamNewPipe/NewPipe/pull/4288#issuecomment-697097581))? But you'll probably want to continue that discussion in #4272 if the issue occurs because of one of avently's commits.
like this
FS-black

**Comment by blackbox87** [comment]
@fareszr What's your phone model?

If your device has hardware buttons then how that looks might actually be correct, but if you've got the navbar hidden via a setting then that's likely the issue as Avently's version doesn't check the navbar heights or modes.

If I use Avently's version on my device running LineageOS 16 it looks correct.

Screenshot_NewPipe_SmallFixes2

**Comment by fareszr** [comment]
> @fareszr What's your phone model?
> 
> If your device has hardware buttons then how that looks might actually be correct, but if you've got the navbar hidden via a setting then that's likely the issue as Avently's version doesn't check the navbar heights or modes.
> 
> If I use Avently's version on my device running LineageOS 16 it looks correct.
> 
> Screenshot_NewPipe_SmallFixes2

i chose the zoom option to fill the screen.
on you version it fills the entire screen 
but using #4272 it has the same issue from source.

device: S10+ with modpunk's Los 17.1

**Comment by avently** [comment]
@blackbox87 
>if you've got the navbar hidden via a setting then that's likely the issue as Avently's version doesn't check the navbar heights or modes.
I don't need it because Android adjust view bounds for me. On screenshot it looks like he should see nav bar but it is invisible (or only layout was adjusted without showing buttons from SystemUI which is strange because others don't have such problem).




@fareszr 
- what apk did you install from my PRs? Give me a link, please
- could you upload a screenshot from this PR where you see a correct result in the same situation I want to compare working vs non-working look of the app
- do you have any specific settings in your rom related to navigation bar?
- what's navigation bar type you have? Three buttons, two buttons, one button, no buttons at all?

And lets move to #4272 since that pr is a better place for such discussion

**Comment by avently** [comment]
@fareszr it's a known bug of your ROM
https://forum.xda-developers.com/showpost.php?p=82170539&postcount=3

Known bugs ->
>in full screen apps there may be a black bar to hide the camera cutout which can not be disabled 

So, will not be fixed, ask your rom developer

**Comment by blackbox87** [comment]
@fareszr Oops, sorry. I see that you mentioned the phone model in your original post.

Screenshot_NewPipe_SmallFixes2_2

For me it still works if I select "zoom" or "fit", but I'm using Android 9. And you can see that it's added padding to the left so that the controls avoid my phones notch and to the right so that there's space for the navbar.

I think the issue might be caused by one of the fitsSystemWindows lines that @avently removed. That's a total guess though, so you'll want to post to #4272 to help diagnose the issue.

@avently I thought that at first, but his device has a cutout at the top of the screen and there's no padding on the right side to avoid it. That's why I asked if he's using a setting to hide the navbar.

**Comment by blackbox87** [comment]
@fareszr Oops, sorry. I see that you mentioned the phone model in your original post.

@avently Good catch. And I agree with not trying to fix bugs that are caused by unoffcial ROMs.

I'm not entirely sure why my version still works though, unless it has something to do with the layout changes you've made (removal of fitsSystemWindows). Everything else should be nearly the same, since we both only apply padding to the controls.

**Comment by avently** [comment]
@blackbox87 

In your code you still have FLAG_LAYOUT_NO_LIMITS because you arrange the views manually. I don't do that so I don't have such flag and I let the system do all adjustments. His rom doesn't apply correct adjustments if flag FULLSCREEN is set

**Comment by blackbox87** [comment]
@avently Ah! Of course, that'll do it.

I do agree with not fixing the issue in this case. I still don't agree on the other issue though, since almost every Android 10 and 11 ROM that isn't based on pure AOSP has the option to hide the gesture bar. But if someone wants to fix it then I've shared a little code that does the trick. I personally don't have the time to do it and I don't want to keep rebasing the code on other peoples PRs.

**Comment by fareszr** [comment]
well thats unfortunate i am stuck with non-unified player now, all other apps work correctly including youtube its only an issue with newpipe debug edition and avently PR.

**Comment by avently** [comment]
@fareszr that's interesting, so current release works ok for you?
Let's find out why. Unified and non-unified should work the same in case of navigation bar. I'll post some test apks for you in #4272.
Answer on these questions inside #4272:

-what apk did you install from my PRs? Give me a link, please
-could you upload a screenshot from current released version 0.19.8 where you see a correct result in the same situation? I want to compare working vs non-working look of the app
- do you have any specific settings in your rom related to navigation bar?
-what's navigation bar type do you have? Three buttons, two buttons, one button, no buttons at all?

**Comment by fareszr** [comment]
> @fareszr that's interesting, so current release works ok for you?
> Let's find out why. Unified and non-unified should work the same in case of navigation bar. I'll post some test apks for you in #4272.
> Answer on these questions inside #4272:
> 
>     * what apk did you install from my PRs? Give me a link, please
> 
>     * could you upload a screenshot from current released version 0.19.8 where you see a correct result in the same situation? I want to compare working vs non-working look of the app
> 
>     * do you have any specific settings in your rom related to navigation bar?
> 
>     * what's navigation bar type do you have? Three buttons, two buttons, one button, no buttons at all?

oh wow its fixed in the final2.zip file, i downloaded the initial test app from the main PR comment.
Thanks and sorry for wasting your time, it would have been better to update the description with the new links!.

**Comment by avently** [comment]
Glad there is no issue to fix:). Testing apk is a thing that changes often so when you test something from a PR take a look at the end of the page. The truth is always at the end:)

**Pull request closed without merging**